Define conversion mapping from OTel Exponential Histograms to Prometheus Native Histograms. (#3079)
open-telemetry/opentelemetry-specification: specification/compatibility/prometheus_and_openmetrics.md

OpenTelemetry Histograms with Delta aggregation temporality SHOULD be aggregated into a Cumulative aggregation temporality and follow the logic above, or MUST be dropped.

### Exponential Histograms

An [OpenTelemetry Histogram](../metrics/data-model.md#exponentialhistogram) with

> a cumulative aggregation temporality MUST be converted to a Prometheus Native

Reviewer: I assume we drop delta temporality exponential histograms.  Can we call that out explicitly?

> a cumulative aggregation temporality MUST be converted to a Prometheus Native

Author: No detailed documentation yet since it is still experimental, see https://prometheus.io/docs/concepts/metric_types/#histogram. There is a [design doc](https://docs.google.com/document/d/1cLNv3aufPZb3fNfaJgdaRBZsInZKKIHo9E6HinJVbpM/edit) for it, do you think it makes sense to link it till the doc is published?

> a cumulative aggregation temporality MUST be converted to a Prometheus Native

Reviewer: probably fine to wait until it is published somewhere

OpenTelemetry Histograms with Delta aggregation temporality SHOULD be aggregated into a Cumulative aggregation temporality and follow the logic above, or MUST be dropped.

### Exponential Histograms

> An [OpenTelemetry Histogram](../metrics/data-model.md#exponentialhistogram) with

Reviewer 2: ```suggestion
An [OpenTelemetry Exponential Histogram](../metrics/data-model.md#exponentialhistogram) with
```

a cumulative aggregation temporality MUST be converted to a Prometheus Native
Histogram as follows:

- `Scale` is converted to the Native Histogram `Schema`. Currently,
  [valid values](https://github.com/prometheus/prometheus/commit/d9d51c565c622cdc7d626d3e7569652bc28abe15#diff-bdaf80ebc5fa26365f45db53435b960ce623ea6f86747fb8870ad1abc355f64fR76-R83)

>   for `schema` are -4 <= n <= 8. Exponential Histogram data points are dropped,
>   if their `Schema` is not within this range.

Reviewer 2: ```suggestion
  for `schema` are -4 <= n <= 8. Exponential Histogram data points are dropped
  if their `Schema` is not within this range.
```

- `Count` is converted to Native Histogram `Count` if the `NoRecordedValue`
  flag is set to `false`.
- `Sum` is converted to the Native Histogram `Sum` if `Sum` is set and the

>   `NoRecordedValue` flag is set to `false`.

Reviewer 2: Do these values need to be set to the Stale NaN value if `NoRecordedValue` is set?

Author: I'll document that, I've done that in the implementation https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17370/files#diff-f3433990c0d97b1edd024a526e8c2a70fd4eea41ba8202615b2e79d91e87db74R98-R100

- `TimeUnixNano` is converted to the Native Histogram `Timestamp` after

>   converting nano to milliseconds.

Reviewer 2: ```suggestion
  converting nanoseconds to milliseconds.
```

- `ZeroCount` is converted directly to the Native Histogram `ZeroCount`.
- `ZeroThreshold`, if set, is converted to the Native Histogram `ZeroThreshold`.
  Otherwise, it is set to the default value `1e-128`.

- The dense bucket layout represented by `Positive` bucket counts and `Offset` is
  converted to the Native Histogram sparse layout represented by `PositiveSpans`
  and `PositiveDeltas`. The same holds for the `Negative` bucket counts
  and `Offset`.
- `Min` and `Max` are not used.
- `StartTimeUnixNano` is not used.

> [OpenTelemetry Histogram](../metrics/data-model.md#exponentialhistogram) metric
> with the delta aggregation temporality is dropped.

Reviewer 2: ```suggestion
[OpenTelemetry Exponential Histogram](../metrics/data-model.md#exponentialhistogram) metrics
with the delta aggregation temporality are dropped.
```

  for `schema` are -4 <= n <= 8. Exponential Histogram data points are dropped
  if their `Schema` is not within this range.
- `Count` is converted to Native Histogram `Count` if the `NoRecordedValue`
  flag is set to `false`, otherwise, Native Histogram `Count` is set to the
  Stale NaN value.
- `Sum` is converted to the Native Histogram `Sum` if `Sum` is set and the
  `NoRecordedValue` flag is set to `false`, otherwise, Native Histogram `Sum` is
  set to the Stale NaN value.
- `TimeUnixNano` is converted to the Native Histogram `Timestamp` after
  converting nanoseconds to milliseconds.
- `ZeroCount` is converted directly to the Native Histogram `ZeroCount`.
- `ZeroThreshold`, if set, is converted to the Native Histogram `ZeroThreshold`.
  Otherwise, it is set to the default value `1e-128`.
- The dense bucket layout represented by `Positive` bucket counts and `Offset` is
  converted to the Native Histogram sparse layout represented by `PositiveSpans`
  and `PositiveDeltas`. The same holds for the `Negative` bucket counts

>   and `Offset`.

Reviewer 3: I think it would be worth adding something to say that Prometheus histograms are indexed by upper boundary while OTel histograms are indexed by lower boundary, the result being that the Offset fields are different-by-one.

OpenTelemetry Histograms with Delta aggregation temporality SHOULD be aggregated into a Cumulative aggregation temporality and follow the logic above, or MUST be dropped.

### Exponential Histograms

An [OpenTelemetry Exponential Histogram](../metrics/data-model.md#exponentialhistogram) with
a cumulative aggregation temporality MUST be converted to a Prometheus Native
Histogram as follows:

> - `Scale` is converted to the Native Histogram `Schema`. Currently,

Reviewer 4: Should we, instead, downscale where appropriate?

e.g. we could move from scale > 8 to scale 8

Author: That makes sense. I'll add your suggestion now.

  [valid values](https://github.com/prometheus/prometheus/commit/d9d51c565c622cdc7d626d3e7569652bc28abe15#diff-bdaf80ebc5fa26365f45db53435b960ce623ea6f86747fb8870ad1abc355f64fR76-R83)
  for `schema` are -4 <= n <= 8. Exponential Histogram data points are dropped

>   if their `Schema` is not within this range.

Reviewer 5: This seems like more reason to support [MaxScale](https://github.com/open-telemetry/opentelemetry-specification/pull/3017), and perhaps a min scale. Though its less obvious what do it if you reach a min scale and the measurements don't fit in the available buckets.

> - `Count` is converted to Native Histogram `Count` if the `NoRecordedValue`

Reviewer 5: What is `NoRecordedValue`?

Author: It is a data point flag signaling a missing value in series, see https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L324-L327.

I couldn't find any documentation on data point flags. I'll double-check that and open an issue if I wouldn't able to find anything.

Reviewer 4: @kovrus / @jack-berg I think this was added as a protocol-only option for prometheus compatibility (i.e. ingesting prometheus).  We don't expose it in our SDKs (normally) because we don't expect them to create these types of points.

Author: Created an issue for that https://github.com/open-telemetry/opentelemetry-specification/issues/3096

OpenTelemetry Histograms with Delta aggregation temporality SHOULD be aggregated into a Cumulative aggregation temporality and follow the logic above, or MUST be dropped.

### Exponential Histograms

An [OpenTelemetry Exponential Histogram](../metrics/data-model.md#exponentialhistogram) with
a cumulative aggregation temporality MUST be converted to a Prometheus Native
Histogram as follows:

- `Scale` is converted to the Native Histogram `Schema`. Currently,
  [valid values](https://github.com/prometheus/prometheus/commit/d9d51c565c622cdc7d626d3e7569652bc28abe15#diff-bdaf80ebc5fa26365f45db53435b960ce623ea6f86747fb8870ad1abc355f64fR76-R83)
  for `schema` are -4 <= n <= 8.

>   If `Scale` is > 8 then Exponential Histogram data points SHOULD be downscaled

Reviewer 6: Do folks think it would be useful to specify that OTel SDK Prometheus exporters should configure [MaxScale](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#base2-exponential-bucket-histogram-aggregation) param so they don't need to implement downscaling at all?

Reviewer 6: I see Jack already left a similar comment

Reviewer 5: Still need to figure out what they should do if the scale is less than -4...

Author: @jack-berg @jsuereth Would it be acceptable to proceed with this specification change as it is and handle max / min scale separately? This part of the specification is marked as experimental, so I assume it is subject to some future changes.
